odl/solvers/nonsmooth/difference_convex.py

```diff
@@ -0,0 +1,243 @@
+    niter : int
+        Number of iterations.
+    callback : callable, optional
+        Function called with the current iterate after each iteration.
+
+    Notes
+    -----
+    The algorithm is described in Section 3 and in particular in Theorem 3 of
+    `[TA1997] <http://journals.math.ac.vn/acta/pdf/9701289.pdf>`_. The problem
+
+    .. math::
+        \min g(x) - h(x)
+
+    has the first-order optimality condition :math:`0 \in \partial g(x) -
+    \partial h(x)`, i.e., aims at finding an :math:`x` so that there exists a
+    common element
+
+    .. math::
+        y \in \partial g(x) \cap \partial h(x).
+
+    The element :math:`y` can be seen as a solution of the Toland dual problem
+
+    .. math::
+        \min h^*(y) - g^*(y)
+
+    and the iteration is given by
+
+    .. math::
+        y_n \in \partial h(x_n), \qquad x_{n+1} \in \partial g^*(y_n)
+
+    for :math:`n\geq 0`.
+
+    References
+    ----------
+    [TA1997] Tao, P D, and An, L T H. *Convex analysis approach to d.c.
+    programming: Theory, algorithms and applications*. Acta Mathematica
+    Vietnamica, 22.1 (1997), pp 289--355.
+
+    See also
+    --------
+    """
+#    `prox_dca`, `doubleprox_dc`
+#    """
+    space = g.domain
+    if h.domain != space:
+        raise ValueError('`g.domain` and `h.domain` need to be equal, but '
+                         '{} != {}'.format(space, h.domain))
+    for _ in range(niter):
+        g.convex_conj.gradient(h.gradient(x), out=x)
+
+        if callback is not None:
+            callback(x)
+
+
+def prox_dca(x, g, h, niter, gamma, callback=None):
+    """Proximal DCA of Sun, Sampaio and Candido.
+
+    This algorithm solves a problem of the form ::
+
+        min_x g(x) - h(x)
+
+    where ``g`` and ``h`` are two proper, convex and lower semicontinuous
+    functions.
+
+    Parameters
+    ----------
+    x : `LinearSpaceElement`
+        Initial point, updated in-place.
+    g : `Functional`
+        Convex part. Needs to provide a `Functional.convex_conj` with a
+        `Functional.proximal` factory.
+    h : `Functional`
+        The negative of the concave part. Needs to provide a
+        `Functional.proximal` factory.
+    niter : int
+        Number of iterations.
+    gamma : positive float
+        Stepsize of the proximal steps.
+    callback : callable, optional
+        Function called with the current iterate after each iteration.
+
+    Notes
+    -----
+    The algorithm was proposed as Algorithm 2.3 in
+    `[SSC2003] <http://www.global-sci.org/jcm/readabs.php?vol=21&\
+no=4&page=451&year=2003&ppage=462>`_. It solves the problem
+
+    .. math ::
+        \\min g(x) - h(x)
+
+    by involving subgradients of :math:`h` and proximal points of :math:`g^*`.
+    The iteration is given by
+
+    .. math ::
+        y_n \\in \\partial h(x_n), \\qquad x_{n+1}
+            = \\mathrm{Prox}^{\\gamma}_g(x_n + \\gamma y_n)
+
+    Here, :math:`\\gamma` is the stepsize parameter ``gamma``.
+
```

Comment:
Perhaps a comment on the difference to the "simple" DCA? I guess the main advantage is that you don't need the subgradient of `g`, but only its `prox`.